nano/node/online_reps.cpp

```diff
@@ -1,101 +1,131 @@
+#include <nano/node/nodeconfig.hpp>
 #include <nano/node/online_reps.hpp>
 #include <nano/secure/blockstore.hpp>
-#include <nano/secure/common.hpp>
 #include <nano/secure/ledger.hpp>
 
-nano::online_reps::online_reps (nano::ledger & ledger_a, nano::network_params & network_params_a, nano::uint128_t minimum_a) :
-ledger (ledger_a),
-network_params (network_params_a),
-minimum (minimum_a)
+nano::online_reps::online_reps (nano::ledger & ledger_a, nano::node_config const & config_a) :
+ledger{ ledger_a },
+config{ config_a }
 {
 	if (!ledger.store.init_error ())
 	{
 		auto transaction (ledger.store.tx_begin_read ());
-		online = trend (transaction);
+		trended_m = calculate_trend (transaction);
 	}
 }
 
 void nano::online_reps::observe (nano::account const & rep_a)
 {
 	if (ledger.weight (rep_a) > 0)
 	{
+		auto now = std::chrono::steady_clock::now ();
 		nano::lock_guard<std::mutex> lock (mutex);
-		reps.insert (rep_a);
+		auto new_insert = reps.get<tag_account> ().erase (rep_a) == 0;
+		reps.insert ({ now, rep_a });
+		auto cutoff = reps.get<tag_time> ().lower_bound (now - std::chrono::seconds (config.network_params.node.weight_period));
+		auto trimmed = reps.get<tag_time> ().begin () != cutoff;
+		reps.get<tag_time> ().erase (reps.get<tag_time> ().begin (), cutoff);
+		if (new_insert || trimmed)
+		{
+			online_m = calculate_online ();
+		}
 	}
 }
 
 void nano::online_reps::sample ()
 {
-	auto transaction (ledger.store.tx_begin_write ({ tables::online_weight }));
-	// Discard oldest entries
-	while (ledger.store.online_weight_count (transaction) >= network_params.node.max_weight_samples)
+	nano::unique_lock<std::mutex> lock (mutex);
+	nano::uint128_t online_l;
```

Comment:
Can be `nano::uint128_t online_l (online_m);` or `nano::uint128_t online_l = online_m;` probably